### PR TITLE
New resource/ data source `azurerm_iothub_digital_twins`

### PR DESCRIPTION
**Pull request by yupwei68**
Fix https://github.com/terraform-providers/terraform-provider-azurerm/issues/9211

=== RUN   TestAccAzureRMDigitalTwins_basic
=== PAUSE TestAccAzureRMDigitalTwins_basic
=== CONT  TestAccAzureRMDigitalTwins_basic
--- PASS: TestAccAzureRMDigitalTwins_basic (170.28s)
=== RUN   TestAccAzureRMDigitalTwins_requiresImport
=== PAUSE TestAccAzureRMDigitalTwins_requiresImport
=== CONT  TestAccAzureRMDigitalTwins_requiresImport
--- PASS: TestAccAzureRMDigitalTwins_requiresImport (147.93s)
=== RUN   TestAccAzureRMDigitalTwins_complete
=== PAUSE TestAccAzureRMDigitalTwins_complete
=== CONT  TestAccAzureRMDigitalTwins_complete
--- PASS: TestAccAzureRMDigitalTwins_complete (170.48s)
=== RUN   TestAccAzureRMDigitalTwins_update
=== PAUSE TestAccAzureRMDigitalTwins_update
=== CONT  TestAccAzureRMDigitalTwins_update
--- PASS: TestAccAzureRMDigitalTwins_update (284.05s)

Process finished with exit code 0

### CODE REVIEW

File: azurerm/internal/services/iothub/validate/digital_twins.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+)
+
+func DigitaltwinsName() schema.SchemaValidateFunc {
+	return validation.StringMatch(
+		regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9-]{1,61}[A-Za-z0-9]$`),
+		`Name contains invalid characters or exceeds allowed length.`,
+	)
+}
```
Comment (WodansSon) [inline azurerm/internal/services/iothub/validate/digital_twins.go:15]: Can we make this function more clear to the end user in case an error does occur? WDYT?
```suggestion
func digitalTwinsName(i interface{}, k string) (warnings []string, errors []error) {
	v, ok := i.(string)
	if !ok {
		errors = append(errors, fmt.Errorf("expected type of %s to be string", k))
		return
	}
	
	if len(v) < 3 {
		errors = append(errors, fmt.Errorf("length should be greater than %d, got %q", 3, v))
		return
	}
	
	if len(v) > 63 {
		errors = append(errors, fmt.Errorf("length should be less than %d, got %q", 63, v))
		return
	}
	
	if !regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9-]+[A-Za-z0-9]$`).MatchString(v) {
		errors = append(errors, fmt.Errorf("expected value of %s does not match regular expression, got %v", k, v))
		return
	}
	return
}
```

File: azurerm/internal/services/iothub/validate/digital_twins_test.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
+package validate
+
+import (
+	"testing"
+)
+
+func TestDigitaltwinsName(t *testing.T) {
```
Comment (WodansSon) [inline azurerm/internal/services/iothub/validate/digital_twins_test.go:7]: Can we add two more test cases here? One for exact length (63 chars) and one for a name that is too long (64 chars)?

**Review by WodansSon — CHANGES_REQUESTED** [review]
@yupwei68 Thank you for this PR, it's looking really good so far but I left a few comments. If you can address those I think this will be good to go! 🚀

**Comment by yupwei68** [comment]
Thanks @WodansSon for your comments. Corresponding changes have been pushed. Please continue reviewing.

File: azurerm/internal/services/iothub/iothub_digital_twins_resource.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/iothub/validate"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tags"
+	azSchema "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmDigitalTwins() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmproperties,
```
Comment (WodansSon) [inline azurerm/internal/services/iothub/iothub_digital_twins_resource.go:24]: I think this one got replaced by accident? Should it be `resourceArmDigitalTwinsCreate`?
```suggestion
		Create: resourceArmDigitalTwinsCreate,
```

File: azurerm/internal/services/iothub/iothub_digital_twins_resource.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+			"host_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"tags": tags.Schema(),
+		},
+	}
+}
+func resourceArmproperties(d *schema.ResourceData, meta interface{}) error {
```
Comment (WodansSon) [inline azurerm/internal/services/iothub/iothub_digital_twins_resource.go:62]: I think this one got replaced by accident? Should it be resourceArmDigitalTwinsCreate?

```suggestion
func resourceArmDigitalTwinsCreates(d *schema.ResourceData, meta interface{}) error {
```

**Review by WodansSon — CHANGES_REQUESTED** [review]
@yupwei68 Almost there, I think there was a search and replace error, but get those two issues fixed up and this resource LGTM! 🚀

**Comment by yupwei68** [comment]
@WodansSon Sorry for the mistakes. Please continue reviewing.

**Review by WodansSon — APPROVED** [review]
@yupwei68 Thanks for pushing those changes, this LGTM now! 🚀

**Comment by WodansSon** [comment]
image

File: azurerm/internal/services/iothub/validate/digital_twins.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+		return
+	}
+
+	if len(v) > 63 {
+		errors = append(errors, fmt.Errorf("length should be less than %d, got %q", 63, v))
+		return
+	}
+
+	if !regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9-]+[A-Za-z0-9]$`).MatchString(v) {
+		errors = append(errors, fmt.Errorf("expected value of %s does not match regular expression, got %v", k, v))
```
Comment (tombuildsstuff) [inline azurerm/internal/services/iothub/validate/digital_twins.go:26]: this error message isn't helpful to users - can we explain the validation requirements here?

File: azurerm/internal/services/iothub/iothub_digital_twins_resource.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+	if resp.ID == nil || *resp.ID == "" {
+		return fmt.Errorf("empty or nil ID returned for Digital Twins %q (Resource Group %q) ID", name, resourceGroup)
+	}
+
+	id, err := parse.DigitalTwinsID(*resp.ID)
+	if err != nil {
+		return err
+	}
+	d.SetId(id.ID(subscriptionId))
```
Comment (tombuildsstuff) [inline azurerm/internal/services/iothub/iothub_digital_twins_resource.go:108]: we can remove all of this in favour of using an ID Formatter directly here?

File: azurerm/internal/services/iothub/iothub_digital_twins_resource.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+		Tags:     tags.Expand(d.Get("tags").(map[string]interface{})),
+	}
+
+	future, err := client.CreateOrUpdate(ctx, resourceGroup, name, properties)
+	if err != nil {
+		return fmt.Errorf("creating Digital Twins %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	if err := future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("waiting on creating future for Digital Twins %q (Resource Group %q): %+v", name, resourceGroup, err)
```
Comment (tombuildsstuff) [inline azurerm/internal/services/iothub/iothub_digital_twins_resource.go:92]: this isn't meaningful to users - shouldn't this be "waiting for creation of Digital Twins.."?

File: azurerm/internal/services/iothub/iothub_digital_twins_resource.go [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	existing, err := client.Get(ctx, resourceGroup, name)
+	if err != nil {
+		if !utils.ResponseWasNotFound(existing.Response) {
+			return fmt.Errorf("checking for present of existing Digital Twins %q (Resource Group %q): %+v", name, resourceGroup, err)
+		}
+	}
+	if existing.ID != nil && *existing.ID != "" {
+		return tf.ImportAsExistsError("azurerm_iothub_digital_twins", *existing.ID)
```
Comment (tombuildsstuff) [inline azurerm/internal/services/iothub/iothub_digital_twins_resource.go:78]: (as below) shouldn't this be using the ID Formatter to ensure it's consistent?

File: azurerm/internal/services/iothub/tests/iothub_digital_twins_resource_test.go [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+}
+
+func testAccAzureRMDigitalTwins_template(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-DigitalTwins-%d"
```
Comment (katbyte) [inline azurerm/internal/services/iothub/tests/iothub_digital_twins_resource_test.go:166]: this is in the iohub service ? so should be `acctestRG-iothub-%d`

File: azurerm/internal/services/iothub/client/client.go [hunk 10]
```diff
@@ -1,6 +1,7 @@
 package client
 
 import (
+	"github.com/Azure/azure-sdk-for-go/services/digitaltwins/mgmt/2020-10-31/digitaltwins"
```
Comment (katbyte) [inline azurerm/internal/services/iothub/client/client.go:4]: Actually why was this added to the iothub service? this is a new package import so shouldn't it be a new service package??

File: website/docs/r/iothub_digital_twins.html.markdown [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+subcategory: "IoT Hub"
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_iothub_digital_twins"
+description: |-
+  Manages a Digital Twins.
+---
+
+# azurerm_iothub_digital_twins
+
+Manages a Digital Twins.
```
Comment (katbyte) [inline website/docs/r/iothub_digital_twins.html.markdown:11]: Digital Twin what? instance? something else?

File: azurerm/internal/services/iothub/registration.go [hunk 12]
```diff
@@ -21,6 +21,7 @@ func (r Registration) WebsiteCategories() []string {
 // SupportedDataSources returns the supported Data Sources supported by this Service
 func (r Registration) SupportedDataSources() map[string]*schema.Resource {
 	return map[string]*schema.Resource{
+		"azurerm_iothub_digital_twins":            dataSourceDigitalTwins(),
```
Comment (katbyte) [inline azurerm/internal/services/iothub/registration.go:24]: THis is a digital twin what? or will there not be any other resources for digital twins? `azurerm_digital_twins_??????` instance?

File: azurerm/internal/services/iothub/validate/digital_twins.go (hunk 5 above)
Comment (WodansSon) [reply azurerm/internal/services/iothub/validate/digital_twins.go:26]: Fixed in PR #9430

File: azurerm/internal/services/iothub/iothub_digital_twins_resource.go (hunk 6 above)
Comment (WodansSon) [reply azurerm/internal/services/iothub/iothub_digital_twins_resource.go:108]: Fixed in PR #9430

File: azurerm/internal/services/iothub/iothub_digital_twins_resource.go (hunk 7 above)
Comment (WodansSon) [reply azurerm/internal/services/iothub/iothub_digital_twins_resource.go:92]: Fixed in PR #9430

File: website/docs/r/iothub_digital_twins.html.markdown (hunk 11 above)
Comment (WodansSon) [reply website/docs/r/iothub_digital_twins.html.markdown:11]: Fixed in PR #9430

File: azurerm/internal/services/iothub/registration.go (hunk 12 above)
Comment (WodansSon) [reply azurerm/internal/services/iothub/registration.go:24]: Fixed in PR #9430
